### PR TITLE
Cleanup Packages: Move libovsdbops to libovsdb/ops, add libovsdb/util

### PR DESCRIPTION
**Pull request by tssurya**
**- What this PR does and why is it needed**
Currently acl utilities are part of
ovn pkg. We have ANP feature that is
being deleloped in its own pkg
admin_network_policy. It would be nice
to re-use some of the existing ACL utilities and have a common place to keep then without running
into cyclic dependencies on imports.

See https://github.com/ovn-org/ovn-kubernetes/pull/3659#discussion_r1270511328 and https://github.com/ovn-org/ovn-kubernetes/pull/3659#discussion_r1257988920 for more details.

**- Special notes for reviewers**
No functional changes, this will help with https://github.com/ovn-org/ovn-kubernetes/pull/3659

**- How to verify it**

**- Description for the changelog**
`Move ovn/acl.go into libovsdb/util/acl.go`

### CODE REVIEW

**Review by npinaeva — APPROVED** [review]
nice!

**Comment by jcaamano** [comment]
hmm not really convinced on how this looks now that I see it, having the libovsdbops import in util, which would mean if we ever need to use util from libovsdbops, which is not unreasonable, we will have a circular dep.

Ideally, util should not import any other ovn-k package except types and generated packages. It already imports config package which has given us a bit of a headache in the past.

Maybe we can leave the methods that rely on  libovsdbops out of it?

**Comment by tssurya** [comment]
> hmm not really convinced on how this looks now that I see it, having the libovsdbops import in util, which would mean if we ever need to use util from libovsdbops, which is not unreasonable, we will have a circular dep.
> 
> Ideally, util should not import any other ovn-k package except types and generated packages. It already imports config package which has given us a bit of a headache in the past.
> 
> Maybe we can leave the methods that rely on libovsdbops out of it?

I see value in combining the logging functions into the util because when I want to do ANP logging I am going to reuse them. If that stays in ovn package then again I will have circular dependency... Example how do I use `BuildACL` stuff if that is under ovn package?
I think its the other way around, libovsdbops should not import util - everybody else should import libovsdbops haha

jokes apart, I don't know how to fix that if we don't put it in a new package...

**Comment by tssurya** [comment]
If I were to create a new package util/acl instead of keeping it under package util will that work better similar to batching? I don't know how that will look if now libovsdbops wants to import util but util.acl is the one importing libovsdbops

If not then I have to just create an entirely new package called util_acl or something outside the util package

**Comment by jcaamano** [comment]
>  I think its the other way around, libovsdbops should not import util - everybody else should import libovsdbops haha

util has very basic functions like stringers :p

> If I were to create a new package util/acl instead of keeping it under package util will that work better similar to batching?

Perhaps not the best layout but I think it would be better, so go for it

**Comment by npinaeva** [comment]
I would also like to have better package boundaries, but libovsdbops is already imported from the util
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/util.go#L17
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/ovn.go#L17
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/net.go#L14
so we are not making it much worse :P

**Comment by jcaamano** [comment]
> I would also like to have better package boundaries, but libovsdbops is already imported from the util https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/util.go#L17 https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/ovn.go#L17 https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/net.go#L14 so we are not making it much worse :P

:/

I guess this might be a good time to decide a way forward to change those as well?

**Comment by tssurya** [comment]
> > I would also like to have better package boundaries, but libovsdbops is already imported from the util https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/util.go#L17 https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/ovn.go#L17 https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/net.go#L14 so we are not making it much worse :P

haha thanks for pointing this out, I had no idea we were already doing this..

> 
> :/
> 
> I guess this might be a good time to decide a way forward to change those as well?

hmm so let's see, we have https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/net.go#L14
and the functions `ExtractPortAddresses`, `GetLRPAddrs` and `GetOVSPortMACAddress` don't really belong to net.go 

second we have: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/ovn.go#L17
and again `CreateMACBinding` this guy is using libovsdbops, 

third we have https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/util.go#L17 where `GetNBZone` and `UpdateNodeSwitchExcludeIPs` are using libovsdbops.

I would just convert the file `https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/ovn.go` into an ovn_util package within the util package or something and move the `ExtractPortAddresses`, `GetLRPAddrs` and `GetOVSPortMACAddress`, `CreateMACBinding`, `GetNBZone`  and `UpdateNodeSwitchExcludeIPs` also into this file.. I can find a better place for `PlatformTypeIsEgressIPCloudProvider`...

WDYBT?

**Comment by jcaamano** [comment]
> > > I would also like to have better package boundaries, but libovsdbops is already imported from the util https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/util.go#L17 https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/ovn.go#L17 https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/net.go#L14 so we are not making it much worse :P
> 
> haha thanks for pointing this out, I had no idea we were already doing this..
> 
> > :/
> > I guess this might be a good time to decide a way forward to change those as well?
> 
> hmm so let's see, we have https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/net.go#L14 and the functions `ExtractPortAddresses`, `GetLRPAddrs` and `GetOVSPortMACAddress` don't really belong to net.go
> 
> second we have: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/ovn.go#L17 and again `CreateMACBinding` this guy is using libovsdbops,
> 
> third we have https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/util.go#L17 where `GetNBZone` and `UpdateNodeSwitchExcludeIPs` are using libovsdbops.
> 
> I would just convert the file `https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/ovn.go` into an ovn_util package within the util package or something and move the `ExtractPortAddresses`, `GetLRPAddrs` and `GetOVSPortMACAddress`, `CreateMACBinding`, `GetNBZone` and `UpdateNodeSwitchExcludeIPs` also into this file.. I can find a better place for `PlatformTypeIsEgressIPCloudProvider`...
> 
> WDYBT?

Perhaps we can move the original `libovsdbops` package to `libovsdb/ops` and then have a `libovsdb/util` with these things thrown in there distributed as `libovsdb/util/<name>.go`?

**Comment by tssurya** [comment]
> 
> Perhaps we can move the original `libovsdbops` package to `libovsdb/ops` and then have a `libovsdb/util` with these things thrown in there distributed as `libovsdb/util/<name>.go`?

not a bad idea.. let me give this a try to make sure I don't run into other dependency issues..

**Comment by tssurya** [comment]
> > Perhaps we can move the original `libovsdbops` package to `libovsdb/ops` and then have a `libovsdb/util` with these things thrown in there distributed as `libovsdb/util/<name>.go`?
> 
> not a bad idea.. let me give this a try to make sure I don't run into other dependency issues..

this turned out fine.. it was a bit painful to get all the parts of code converted, but it was worth it. I like the layout better now and it makes much more sense for these functions to go inside libovsdb/util rather than the main util.

File: go-controller/pkg/libovsdb/util/logical_switch_port.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+		if ip == nil {
+			return nil, nil, fmt.Errorf("failed to parse logical switch port %q IP %q is not a valid ip address", lsp.Name, addr)
+		}
+		ips = append(ips, ip)
+	}
+	return mac, ips, nil
+}
+
+// GetLRPAddrs returns the addresses for the given logical router port
+func GetLRPAddrs(nbClient client.Client, portName string) ([]*net.IPNet, error) {
```
Comment (tssurya) [inline go-controller/pkg/libovsdb/util/logical_switch_port.go:47]: oops should I call this maybe port.go ? to make it generic...?

Comment (tssurya) [reply go-controller/pkg/libovsdb/util/logical_switch_port.go]: done.

File: go-controller/pkg/libovsdb/util/acl.go [hunk 2]
```diff
@@ -136,6 +136,12 @@ func getACLMatch(portGroupName, match string, aclDir aclDirection) string {
 	return aclMatch
 }
 
+// ACL logging severity levels
```
Comment (flavio-fernandes) [inline go-controller/pkg/libovsdb/util/acl.go:139]: note to self: moved here from go-controller/pkg/ovn/ovn.go

File: go-controller/pkg/ovn/policy_test.go [hunk 3]
```diff
@@ -18,6 +18,7 @@ import (
 	libovsdbclient "github.com/ovn-org/libovsdb/client"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
 	libovsdbops "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/libovsdb/ops"
+	libovsdbutil "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/libovsdb/util"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
```
Comment (flavio-fernandes) [inline go-controller/pkg/ovn/policy_test.go:22]: I did not look too deep, so I will just ask instead. ;)
No tests here could/should be moved into `libovsdb/util/acl_test.go` ?

**Review by flavio-fernandes — COMMENTED** [review]
looks good to me, except for one nit question on whether we should need `act_test.go`.

thank you for breaking up the changes into multiple commits; made review a lot easier!!!

File: go-controller/pkg/libovsdb/ops/model_client_test.go [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
+package ops
 
 import (
 	"fmt"
-	"github.com/onsi/ginkgo"
 	"reflect"
 	"sort"
 	"testing"
 
+	"github.com/onsi/ginkgo"
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/libovsdb/ops/model_client_test.go:9]: yes not me

File: go-controller/pkg/ovn/address_set/fake_address_set.go [hunk 5]
```diff
@@ -2,14 +2,15 @@ package addressset
 
 import (
 	"fmt"
-	"github.com/onsi/gomega"
 	"net"
 	"sync"
 	"sync/atomic"
 
+	"github.com/onsi/gomega"
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/ovn/address_set/fake_address_set.go:9]: yes not me

File: go-controller/pkg/util/net.go [hunk 6]
```diff
@@ -10,68 +10,11 @@ import (
...
-			return nil, fmt.Errorf("unable to parse network CIDR: %s for router port: %s, err: %v ", network, portName, err)
-		}
-		gwLRPIPs = append(gwLRPIPs, &net.IPNet{
-			IP:   ip,
-			Mask: network.Mask,
-		})
-	}
-	return gwLRPIPs, nil
-}
+var ErrorNoIP = errors.New("no IP available")
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/util/net.go:17]: it was staticcheck complaining... :/ so fixed that as well.

File: go-controller/pkg/libovsdb/util/acl.go (hunk 2 above)
Comment (tssurya) [reply go-controller/pkg/libovsdb/util/acl.go:139]: yup, was better to move it here..

File: go-controller/pkg/ovn/policy_test.go (hunk 3 above)
Comment (tssurya) [reply go-controller/pkg/ovn/policy_test.go:22]: yeah this file did not have any tests, so we are good for now... nothing extra to move..

**Review by npinaeva — APPROVED** [review]

**Review by jcaamano — APPROVED** [review]

**Pull request merged**